### ISSUE TITLE
da.percentile and np.percentile mismatch

### ISSUE THREAD

**Issue by rzu512** [dask/dask#3099]
Test:
```python
import numpy as np
import dask.array as da
x = da.arange(12, chunks=2)
p = da.percentile(x, 50)
print("dask result", p.compute())
x = x.compute()
p = np.percentile(x, 50)
print("numpy result", p)
```

Result:
```
dask result [10.5]
numpy result 5.5
```

Versions:
```
Python 3.6.3
Dask 0.16.1 installed by pip3 install dask[complete] --user
numpy 1.14.0
```

**Comment by jcrist** [comment]
This is because dask's `percentile` algorithm is approximate, while numpy's is exact. There is no efficient way to do exact percentiles in a parallel/out-of-core manner.

The algorithm used in dask isn't always the best though, a better one (tdigest) is implemented in https://github.com/jcrist/crick. I've been meaning to port over dask to use this one as well, but haven't found the time (PRs welcome).

A quick implementation using `dask.delayed`:

```python
In [1]: %paste
import numpy as np
import dask.array as da
x = da.arange(12, chunks=2)
p = da.percentile(x, 50)

## -- End pasted text --

In [2]: import crick

In [3]: def chunk(x):
   ...:     t = crick.TDigest()
   ...:     t.update(x)
   ...:     return t
   ...:

In [4]: def merge(ts):
   ...:     t = crick.TDigest()
   ...:     t.merge(*ts)
   ...:     return t
   ...:

In [5]: from dask import delayed

In [6]: def tdigest(x):
   ...:     chunks = [delayed(chunk, pure=True)(b) for b in x.to_delayed()]
   ...:     return delayed(merge, pure=True)(chunks)
   ...:

In [7]: t = tdigest(x)

In [8]: t2 = t.compute()

In [9]: t2.quantile(0.5)
Out[9]: 5.5

In [10]: p.compute()
Out[10]: array([ 10.5])
```

Related to #731 and #1225.

**Comment by rzu512** [comment]
When does the current approximation work?

I would expect it works for large arrays with small chunks.
But, I get the following even if I change the 3rd line to `x = da.arange(1200, chunks=10)`.
```
dask result [1149.5]
numpy result 599.5
```
Is it always off by a factor of `50 / 100`?

**Comment by rzu512** [comment]
Looks like it is off by the factor of `q / 100`:

Test:
```python
import numpy as np
import dask.array as da
q = 30
x = da.arange(1200, chunks=10)
p = da.percentile(x, q) * q/100
print("dask result", p.compute())
x = x.compute()
p = np.percentile(x, q)
print("numpy result", p)
```
Result:
```
dask result [357.81]
numpy result 359.7
```

**Comment by jcrist** [comment]
Perfectly sorted data like `arange` is a tough one for the current algorithm. Try something a bit more random:

```python
In [8]: x = da.random.gamma(10, 1, size=10000, chunks=1000)

In [9]: da.percentile(x, 50).compute()
Out[9]: array([ 9.89036079])

In [10]: np.percentile(x, 50)
Out[10]: 9.6435231217598307

In [11]: da.percentile(x, 10).compute()
Out[11]: array([ 6.48986224])

In [12]: np.percentile(x, 10)
Out[12]: 6.2448776252782388
```

**Comment by rzu512** [comment]
Thanks. I think I will try other ways now if possible. 
Maybe `da.random.choice(x.size, p=x, size=n_selected)` to sub-sample if it isn't too slow.

**Comment by rzu512** [comment]
I haven't used Cython before.
What is the main difficulty of porting the code at https://github.com/jcrist/crick over Dask?
Can Dask or Cloudpickle handle passing struct with pointer to another struct?

**Comment by jcrist** [comment]
> The java version of the code says the newer versions don't need dynamic allocation after creating the t-digest now

Yes. The algorithm implemented in crick is the same one.

> I haven't used Cython before.
> Can Dask or Cloudpickle handle passing struct with pointer to another struct?

None of this is necessary here. We're not porting the code from crick, we'd be making use of it as an optional dependency. The example I provided above is one functional way of doing that, but if/when we add it to dask we'd want to create the graphs directly rather than using `delayed`.

**Comment by rzu512** [comment]
Hmm. I was thinking that after calculating t-digest of a block in a 1D array, two block merges by merging their t-digest. This would require Dask/Cloud pickle to send the t-digest from one block to another block. The t-digest class contains struct with a pointer to a centroid struct. But I guess probably passing t-digest between blocks wouldn't be a problem.

When I skim the code yesterday, I got the idea that a t-digest instance works for 1D array. I haven't checked if the instance can directly work for multi-dimensional array. If the user asks for calculating percentile on a 2D Dask array A along an axis, do we have to create a t-digest object for each column / row of each block of A?

Anyway, I will first try to make a percentile function that works for 1D arrays in my spare time. That sounds simple enough.

**Comment by jcrist** [comment]
- Crick doesn't handle axis, we'd need 1 tdigest per element in the result.
- TDigest's pickle fine, you shouldn't worry about moving them around.

**Comment by rzu512** [comment]
I found a multi-dimensional weighted histogram for C++ and Python: [github.com/HDembinski/histogram](https://github.com/HDembinski/histogram)
It can add (merge) two histograms together. I haven't figured out the memory bounds or potential limitations for highly non-uniform data yet. Since my project can possibly use the multi-dimensional histogram, I am sorry that I would investigate porting this to Dask first before working on crick.

**Comment by rzu512** [comment]
While t-digest only promises to calculate quantiles on 1-D data, I try to use it to estimate multi-dimensional histogram:

Blobs made by scikit-learn and converted to ints (50000 data points)
![1](https://user-images.githubusercontent.com/32720823/35778539-8cb48e78-098d-11e8-98c4-a82075f870d1.png)
Reconstruction from np.histogram of histogram of ints (bins: 20000)
![2](https://user-images.githubusercontent.com/32720823/35778541-8e386530-098d-11e8-84df-11fef95627a8.png)
Reconstruction from crick's histogram of histogram of ints (compression: 20000, bins: 20000)
![3](https://user-images.githubusercontent.com/32720823/35778542-8f413682-098d-11e8-8677-11e6d9d8d5e4.png)

Since I found I actually only need compressed 3d histogram, I am sorry that I would look for an implementation to port to Dask before working on crick.

```python
from sklearn.datasets.samples_generator import make_blobs
import matplotlib.pyplot as plt
import crick
import numpy as np

def plot(counts, bin_edge, shape0):
	bin_centers = (bin_edge[1:] + bin_edge[:-1])/2
	int_bin_centers = bin_centers.astype(int) 
	a = int_bin_centers // shape0
	b = int_bin_centers % shape0
	plt.scatter(a, b, c=counts)
	plt.show()

X, y = make_blobs(n_samples=50000, centers=8, n_features=4,
                  random_state=0)

resolution = 0.01
X_int = (X / resolution).astype(int)
max_x = np.amax(X_int, axis=0)
min_x = np.amin(X_int, axis=0)
shape_x = max_x - min_x

X_int -= min_x
x_int = X_int[:,0] * shape_x[0] + X_int[:, 1] 

x_int_a = x_int // shape_x[0]
x_int_b = x_int % shape_x[0]
plt.scatter(x_int_a, x_int_b)
plt.show()

# recontruction by np.histogram
bins=20000
counts, bin_edge = np.histogram(x_int, bins=bins)
plot(counts, bin_edge, shape_x[0])

# Recontruction by crick histogram
x = crick.TDigest(compression=20000)
x.update(x_int)
counts, bin_edge = x.histogram(bins=bins)
plot(counts, bin_edge, shape_x[0])
```

**Comment by rzu512** [comment]
I removed the limits of compression in the crick source code.
After that, reconstruction by crick's histogram can work if I use a large compression value.
The compression ratio (size of original data / size of tdigest array) is about 50. And it looks ok.
But this is really not for multi-dimensional histogram... things like [mhist-p](https://link.springer.com/chapter/10.1007/978-3-642-13312-1_15) works better for that purpose.
For now, tdigest seems to work for me.

**Comment by jcrist** [comment]
A few comments:

> Reconstruction by crick's histogram of ints (compression: 20000, bins: 20000)

You seem to have figured this out, but in this case you're generating a 1D histogram for a 2D array. All values for each y value are exactly the same because of this.

> Reconstruction by creating a crick histogram for each row of data.

This is what I'd do if I was to use crick for this.

> But this is really not for multi-dimensional histogram... things like mhist-p works better for that purpose.

Correct. The t-digest algorithm only works for 1D data. This makes sense for dataframes and for streams, but less for ND arrays. I have no immediate plans to implement an ND version.

**Comment by cr458** [comment]
Just checking if this approximation issue for the percentile calculation has been resolved, or if there are any plans to resolve it in the near future?

**Comment by mrocklin** [comment]
> Just checking if this approximation issue for the percentile calculation has been resolved, or if there are any plans to resolve it in the near future?

This would be a great contribution if anyone is interested.  

Closing in favor of https://github.com/dask/dask/issues/1225
